Add support for manual step execution ordering using step names and heuristic branch prediction.

contrains opened the pull request:
This pull request adds support for out-of-order execution of workflow steps by the addition of raisable `GoTo` exceptions as well as heuristic branch prediction for progress estimation.

Steps may now referenced directly from within any workflow step using the new `GoToStep` exception.  This exception will preëmpt the sequential execution of the workflow and instead proceed directly to the referenced step.  There is also a `GoToEnd` exception which immediately completes the workflow.

The addition of out-of-order execution and the potential for multiple execution paths by conditionally following different exceptions cause potential difficulty for the calculation of `percent_completed`, so support was also added for defining `paths` for each step to allow the workflow to heuristically calculate potential paths as the workflow progresses and provide a better estimate of a completion percentage.

More information on both of these new features may be found in the documentation.

prryplatypus commented:
Hi, thanks you for the PR! I'm not sure using labels is the way to go. After some thought, I was wondering what you'd think of this other approach instead:

```py
import random

from ergate import EndWorkflow, FutureStep, GoToEnd, GoToStep, Workflow

workflow = Workflow(unique_name="whatever")

# Follows the default path (would be equivalent to paths=[FutureStep("step_2")])
@workflow.step()
def step_1() -> None:
    print("Hello, I am step 1")

# Follows the default path or goes to the end/completes workflow
@workflow.step(paths=[FutureStep("step_3"), EndWorkflow()])
def step_2() -> None:
    if random.randint(0, 1):
        return
    raise GoToEnd()

# Can either follow the default path (step 4), go to step 5, or complete the workflow
@workflow.step(paths=[FutureStep("step_4"), FutureStep("step_5"), WorkflowEnd()])
def step_3() -> None:
    num = random.randint(1, 2)
    if num == 1:
        return
    if num == 2:
        raise GoToStep(step_5)
    raise GoToEnd()

# Follows the default path (step 5)
@workflow.step()
def step_4() -> None:
    print("Hello, I am step 4")
    return

# Follows the default path (end of workflow)
@workflow.step()
def step_5() -> None:
    print("Hello, I am step 5")
    return
```

When registering the workflow into the app, then we can resolve the `FutureStep`s into actual steps, and error if they don't exist.

Let me know what you think of this approach!

prryplatypus commented:
Also worth noting: we should only allow skipping to steps that are further ahead in the workflow (not ones that are previously defined).

contrains commented:
> Hi, thanks you for the PR!

You're welcome. :p

> I'm not sure using labels is the way to go. After some thought, I was wondering what you'd think of this other approach instead:
>
> ...
> 
> When registering the workflow into the app, then we can resolve the `FutureStep`s into actual steps, and error if they don't exist.
> 
> Let me know what you think of this approach!

That's a great idea actually, to use the function names themselves as the step labels.  This seems to make a lot of sense.  I'll give it contemplation to see if I can think of any drawbacks.

> 
> P.D.: Ideally we'd use the `GoToStep` directly in the `paths` definition, but you can't reference callables/steps that are defined later in the file; hence the use of `FutureStep`.

I'm not sure I understand what the difference is between using `GoToStep` and `FutureStep` in the decorator's arguments, as both are accepting a string input.  Is your thought that by using `FutureStep`, the engine will resolve them all once the workflow is fully loaded to ensure they are all valid?

The current code checks label / step validity at the point when they are called by the workflow, but also when calculating path heuristics (which are intentionally optional for backwards compatibility).  Since the paths hinting is still optional, I believe implementing this method with `FutureStep` would lead to the same effective result.

One benefit I could see to `FutureStep` would be for nomenclature to distinguish the operation (`GoToStep`) from the description thereof (`FutureStep`).  To this end, I would wonder if `GoToEnd` should also have a descriptive counterpart for the paths.

contrains commented:
> Also worth noting: we should only allow skipping to steps that are defined further ahead in the workflow (not ones that are previously defined). Otherwise it opens the doors up to it becoming a mess and being much harder to follow.

I don't see the benefit of this limitation.  There might be legitimate cases for wanting to have a loop of steps in a workflow until some condition is met, which would entail jumping backwards in the sequence.  As just one example of a potential usage.

In theory it could lead to a workflow becoming harder to follow if ill-desisnged, but since using labels (or an alternative as described above) and `GoTo` are strictly optional, a user can choose simply to write their workflow linearly (or with `GoTo` only moving forward, akin to `SkipNSteps`) and thereby avoid the issue entirely, whilst still allowing a different user who would benefit from the ability to step backwards the option to do so.

prryplatypus commented:
> I'm not sure I understand what the difference is between using GoToStep and FutureStep in the decorator's arguments, as both are accepting a string input. Is your thought that by using FutureStep, the engine will resolve them all once the workflow is fully loaded to ensure they are all valid?

`GoToStep` takes a step as an argument directly in my example. `FutureStep` is used to reference steps that will be defined later, but haven't been defined yet at the time of using it (so can't receive their object directly).

>  To this end, I would wonder if GoToEnd should also have a descriptive counterpart for the paths.

I did think about this too, and I do feel like it should be this way. My only concern was people potentially raising the descriptive counterpart instead of the exception, but I guess typecheckers could catch that if the descriptive counterpart isn't a subclass of `Exception`. And good documentation too :)

> I don't see the benefit of this limitation. There might be legitimate cases for wanting to have a loop of steps in a workflow until some condition is met, which would entail jumping backwards in the sequence. As just one example of a potential usage.

At this time I'm not too comfortable wanting to lift this limitation. It doesn't mean I'm fully against it, but I feel like the drawbacks (possible infinite loops, much harder to follow code, etc...) exceed the advantages from my point of view.

prryplatypus commented on ergate/workflow.py:
```diff
@@ -1,35 +1,156 @@
+        all: bool = False,
+        exc: ErgateError | None = None,
+    ) -> list[list[WorkflowPathTypeHint]]:
+        # TODO: better way of determining range for infinite loop detection.
```
You won't have to worry about this if we only allow skipping forward :)

contrains commented:
@prryplatypus I think I've updated this to your liking.  I added guard clauses to limit (arbitrarily) reverse progression from `GoToStep`, and fixed a few subtle bugs.

I think `paths` might be less important with this current implementation, as it primarily was designed to prevent `steps_completed` from exceeding `total_steps` in cases where there was reverse progression, or likewise to prevent the `percent_completed` from decreasing when accessing an earlier step (though I think this might have been eliminated elsewise).

contrains commented:
I've made a few extra changes in PR:

* `GoToStep` now accepts the `WorkflowStep` functions themselves as its argument, rather than the string names of those functions.
* `WorkflowStep` objects now store their index internally.
* Paths are precalculated for workflow steps once at startup rather than for every run.
* `UnknownStepNameError` is renamed to `UnknownStepError`
* `Workflow.get_index_by_step_name` is renamed to `get_step_index_by_name`.

prryplatypus commented on ergate/workflow.py:
```diff
@@ -26,10 +39,110 @@ def __iter__(self) -> Iterator[WorkflowStep]:
+                f'Workflow "{self.unique_name}"'
+            )
+
+    def register(self) -> None:
```
Can we rename this to `finalize`?

contrains replied:
Good idea.  `register` didn't sit right with me, but I didn't have a better name.

contrains commented on docs/docs/basics/defining-workflow-path-hints.md:
```diff
@@ -0,0 +1,156 @@
+This logic may be enabled by specifying the `paths=` kwarg for a given `@workflow.step()` decorator.
+
+* It is not necessary to specify `paths` for workflow steps which only return a value via `return`, thus maintaining full compatibility with existing code.
+* It is suggested to define `paths` for steps which utilise the `SkipNSteps` or `GoToEnd` exceptions. 
+* It is recommended to define `paths` for steps which utilise the `GoToStep` exception, or which have branching logic with multiple potential paths.
```
It was intentionally different.  It was written for the version of `GoToStep` which supported reverse progression, which could have allowed for the number of completed steps to exceed the default total number of steps (`len(workflow)`), resulting in completed percentages exceeding 100%, or potentially triggering a premature completion.  

This is not an issue which affects `SkipNSteps` or `GoToEnd` due to their linear nature, and use of `paths` for those merely increases precision for the progress bar.

contrains commented on ergate/job.py:
```diff
@@ -34,13 +35,26 @@ def mark_running(self, step: WorkflowStep) -> None:
     ) -> None:
-        self.steps_completed = min(self.steps_completed + n, total_steps)
+        self.current_step += n
+        self.steps_completed = min(self.steps_completed + 1, total_steps)
```
Is it possible here, when `paths` aren't defined and `total_steps` defaults to `len(workflow)`, to end up with `self.steps_completed == total_steps` prematurely, thus marking the workflow as complete whilst it is still actually incomplete?

I think if so, this would only affect the version supporting negative progression.

prryplatypus replied:
It wasn't usually possible, no, as this method was only called _after_ a step had completed. With negative progression, however, it appears that could probably be the case.

contrains commented on docs/docs/basics/defining-workflow-path-hints.md:
```diff
@@ -0,0 +1,156 @@
+    print("Hello, I am step 2")
+    raise GoToStep(step_3, retval=2)
+
+@workflow.step(path=[GoToEndPath()])
+def step_5() -> None:
+    print("Hello, I am step 5")
+    raise GoToEnd()
```
It is included intentionally as an example of how a step can be skipped over and never revisited, demonstrating the function of `GoToStep` whilst also illustrating that it's possible that some steps might not be run.

contrains commented on docs/docs/basics/defining-workflow-path-hints.md:
```diff
@@ -0,0 +1,156 @@
+For calculating `percent_completed` and `total_steps`, it will use the longest of the potential branches.
+
+When `step_4` is reached and completed, it will then reevaluate the branches to determine which were not taken, and 
+discard them, leaving only the relevant branch(es) for recalcation and refinement of the `percent_completed` and 
```
This is a typo.  Thanks for finding it!

```suggestion
discard them, leaving only the relevant branch(es) for recalculation and refinement of the `percent_completed` and 
```

contrains commented on docs/docs/basics/defining-workflow-path-hints.md:
```diff
@@ -0,0 +1,156 @@
+
+workflow = Workflow(unique_name="workflow_with_paths_2")
+
+@workflow.step(paths=[GoToStepPath("step_two")])
```
I probably imported it just out of excitement when typing all of the other names.

This step could have `GoToStep(step_two)` removed, but then the decorator would more properly not have an argument at all.

contrains commented on docs/docs/basics/defining-workflow-path-hints.md:
```diff
@@ -0,0 +1,156 @@
+5. `step_6`
+6. `step_7`
+
+Note that the `return` path wasn't explicitly specified for `step_4`, but it is automatically deduced and included from
+the function's return type hinting.  If however a work step function ever intentionally returns `None` as a valid 
+workflow path, but additionally has `paths` defined for other paths, then `NextSteppath()` must be manually included in 
+the list of paths.
```
I'll think about how I can reword / reorder things, sure.
